File: SeekReap-Tier-5-Worker-Pool/worker_pool.py
#!/usr/bin/env python3
import os
import time
import signal
import sys
import psycopg2
from contextlib import closing
from urllib.parse import urlparse

# Graceful shutdown
stop_signal = False
def shutdown_handler(signum, frame):
    global stop_signal
    stop_signal = True
signal.signal(signal.SIGINT, shutdown_handler)
signal.signal(signal.SIGTERM, shutdown_handler)

# Load database URL from environment
DATABASE_URL = os.environ.get("DATABASE_URL")
if not DATABASE_URL:
    print("DATABASE_URL environment variable not set")
    sys.exit(1)

# Parse DATABASE_URL for psycopg2
url = urlparse(DATABASE_URL)
DB_PARAMS = {
    "dbname": url.path[1:],   # remove leading slash
    "user": url.username,
    "password": url.password,
    "host": url.hostname,
    "port": url.port,
    "sslmode": "require"
}

def process_job(job):
    job_id, content_id = job
    print(f"Processing job {job_id} ({content_id})")
    with closing(psycopg2.connect(**DB_PARAMS)) as conn:
        with conn.cursor() as cur:
            # Mark started
            cur.execute(
                "UPDATE job_queue SET status='running', started_at=NOW() WHERE job_id=%s",
                (job_id,)
            )
            conn.commit()

            # Simulate job processing (replace with real logic)
            time.sleep(2)

            # Mark completed
            cur.execute(
                "UPDATE job_queue SET status='completed', completed_at=NOW() WHERE job_id=%s",
                (job_id,)
            )
            conn.commit()
            print(f"Completed job {job_id}")

def main():
    print("Worker pool started")
    while not stop_signal:
        try:
            with closing(psycopg2.connect(**DB_PARAMS)) as conn:
                with conn.cursor() as cur:
                    cur.execute(
                        "SELECT job_id, content_id FROM job_queue WHERE status='pending' ORDER BY job_id ASC LIMIT 1 FOR UPDATE SKIP LOCKED"
                    )
                    job = cur.fetchone()
                    if job:
                        process_job(job)
                    else:
                        time.sleep(1)
        except Exception as e:
            print(f"DB connection lost or error: {e}")
            time.sleep(2)

    print("Worker pool exiting")

if __name__ == "__main__":
    main()

File: SeekReap-Tier-5-Worker-Pool/worker_pool.js
const { Client } = require('pg');
const DATABASE_URL = process.env.DATABASE_URL;
if (!DATABASE_URL) { console.error('DATABASE_URL is not set'); process.exit(1); }

const client = new Client({ connectionString: DATABASE_URL, ssl: { rejectUnauthorized: false } });

async function processJob(job) {
    console.log(`Processing job ${job.job_id}: ${job.content_id}`);
    try {
        await new Promise(r => setTimeout(r, 2000)); // simulate work
        await client.query(
            'UPDATE job_queue SET status=$1, completed_at=NOW() WHERE job_id=$2',
            ['completed', job.job_id]
        );
        console.log(`Job ${job.job_id} completed`);
    } catch (err) {
        console.error(`Job ${job.job_id} failed`, err);
        await client.query(
            'UPDATE job_queue SET status=$1 WHERE job_id=$2',
            ['failed', job.job_id]
        );
    }
}

async function runWorker() {
    await client.connect();
    console.log('Worker connected to DB');
    while (true) {
        const res = await client.query(
            "SELECT job_id, content_id FROM job_queue WHERE status='pending' ORDER BY job_id ASC LIMIT 1"
        );
        if (res.rows.length === 0) { await new Promise(r => setTimeout(r, 2000)); continue; }
        await client.query("UPDATE job_queue SET status='in_progress', started_at=NOW() WHERE job_id=$1", [res.rows[0].job_id]);
        await processJob(res.rows[0]);
    }
}

runWorker().catch(err => { console.error('Worker error', err); process.exit(1); });
